src/box/tuple_hash.cc

```diff
@@ -358,59 +370,94 @@ tuple_hash_key_part(uint32_t *ph1, uint32_t *pcarry, struct tuple *tuple,
 	return tuple_hash_field(ph1, pcarry, &field, part->coll);
 }
 
-template <bool has_optional_parts, bool has_json_paths>
+template <bool has_json_paths>
+const char *
+tuple_get_field(struct tuple *tuple, struct key_part *part,
+		uint32_t multikey_idx)
+{
+	struct tuple_format *format = tuple_format(tuple);
+	const char *tuple_raw = tuple_data(tuple);
+	const uint32_t *field_map = tuple_field_map(tuple);
+	if (has_json_paths) {
+		return tuple_field_raw_by_part(format, tuple_raw, field_map,
+					       part, (int)multikey_idx);
+	} else {
+		return tuple_field_raw(format, tuple_raw, field_map,
+				       part->fieldno);
+	}
+}
+
+template <bool has_optional_parts>
+void
+tuple_hash_part(uint32_t *h, uint32_t *carry, uint32_t *total_size,
+		const char *tuple_end, const char **field, struct coll *coll)
+{
+	if (has_optional_parts && (*field == NULL || *field >= tuple_end)) {
+		*total_size += tuple_hash_null(h, carry);
+	} else {
+		*total_size += tuple_hash_field(h, carry, field, coll);
+	}
+}
+
+template <bool has_optional_parts>
+uint32_t
+tuple_hash_slowpath_multikey(struct tuple *tuple, struct key_def *key_def)
+{
+	uint32_t h = HASH_SEED;
+	uint32_t carry = 0;
+	uint32_t total_size = 0;
+	struct tuple_format *format = tuple_format(tuple);
+	const char *tuple_raw = tuple_data(tuple);
+	const uint32_t *field_map = tuple_field_map(tuple);
+	uint32_t multikey_count = tuple_multikey_count(tuple, key_def);
+	const char *tuple_end = (char *)tuple + tuple_size(tuple);
+	for (uint32_t multikey_idx = 0; multikey_idx < multikey_count;
```

Comment:
I'd expect `tuple_hash()` to take a multikey index and hash only the specified part, not all of them. This would let us support multikey hash indexes.

Also, it'd be nice to have some comments to the new functions and some unit tests.

Reply:
Done.